### DISCUSSION TITLE
Feature: Configurable Module Bundler Selection (Webpack/Rspack/Vite)

### DISCUSSION THREAD

**Discussion by Abdel-Monaam-Aouini** [vercel/next.js#71876] [Ideas]
### Goals

1. Add built-in support for switching between Webpack, Rspack, and Vite without experimental flags
2. rovide a standardized configuration API for all bundlers in next.config.js
3. Ensure consistent performance metrics and debugging experience across bundlers

### Non-Goals

1. Creating custom bundler implementations or modifications
2. Supporting unofficial or community bundlers
3. Maintaining separate codebases for different bundlers

### Background

Currently, Next.js primarily uses Webpack as its default bundler . This creates friction for teams wanting to use alternative bundlers in production, as experimental features may break or change. While there are community solutions and workarounds, having native, stable support for multiple bundlers would benefit the ecosystem.

The main alternatives currently available are:

1. Using experimental flags (unstable)
2. Custom webpack configurations (limited)
3. Community forks (fragmented ecosystem)
4. Manual bundler integration (high maintenance)

### Proposal

Implementation through a new configuration option in next.config.js:

![image](https://github.com/user-attachments/assets/57cb5fe2-ee19-4458-904a-954233ac34a2)

**Comment by Netail** [comment]
NextJS had some custom webpack plugins, which they then have to support for esbuild/rollup (Vite) too. RSpack could work as it's a drop-in replacement for Webpack, but as they are working on their own rust-based bundler, I doubt they'll add support for switching webpack compatible bundlers

**Comment by dmythro** [comment]
+1 on this, at least for Rspack as a drop-in replacement. **Can't wait for Turbopack forever!** Especially when there's a production-ready Rspack (which I use on other projects and it is truly awesome).

Currently, with Next.js 15, production builds are super slow and dev builds with Turbopack don't seem as consistent, and a lot of bugs reported without answers. Please go ahead and at least support Rspack for builds!

**Reply by dmythro** [reply]
For comparison, check Astro and its builds. I created new fresh projects with both — and Astro builds it in an instant, while Next.js production builds are so slow even with almost empty 1-page project!

Next.js is good, but builds seem pretty bad and outdated compared to other frameworks.

**Reply by Netail** [reply]
https://github.com/vercel/next.js/pull/74554 ;)

**Reply by Abdel-Monaam-Aouini** [reply]
@Netail nice 🍡

**Reply by dmythro** [reply]
@Netail seems like it is not a huge deal to support Rspack instead of Webpack, besides all the changes on CI, docs etc. Would be also cool if Next.js would have it all as peer dependencies so users could choose the toolchain without installing Webpack, Turbopack etc.

But unfortunately I see Vercel doesn't really like to allow this kind of changes. Same issue with linters where ESLint is there by default and something like Biome can't be used same way instead of it as of now — the only way is to disable ESLint in a config and use Biome separately without any integration into Next.js builds etc.

**Reply by Netail** [reply]
> seems like it is not a huge deal to support Rspack instead of Webpack

I think it required some close collaboration with the Rspack team for missing drop-in replacement features, not everything where some webpack plugins were dependent on were available in Rspack. So it may doesn't look that much on NextJS' side, but could be quite some development on Rspack's side

**Reply by dmythro** [reply]
As I understood, new native plugins and modules are much faster, but still webpack plugins should work as well?
Ideally, of course, config should preferably use native plugins. They really make a difference, but I didn't actually try sticking with old webpack plugins to see if it works :)
